src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs

```diff
@@ -152,11 +175,15 @@ public bool ShouldAnalyze(ISymbol symbol)
                 return false;
             }
 
-            foreach (var location in symbol.Locations)
+            foreach (var syntaxRef in symbolEvent.DeclaringSyntaxReferences)
             {
-                if (FilterFileOpt.Value.SourceTree == location.SourceTree && ShouldInclude(location.SourceSpan))
+                if (syntaxRef.SyntaxTree == FilterFileOpt.Value.SourceTree)
                 {
-                    return true;
+                    var node = getTopmostNodeForAnalysis(symbolEvent.Symbol, syntaxRef, symbolEvent.Compilation, cancellationToken);
```

Comment:
>getTopmostNodeForAnalysis [](http://example.com/codeflow?start=31&length=25)

Is there a concern around materializing syntax?

Reply:
This should be fine here given we would already walk through all the symbol's declaring references to run analyzers on the nodes. Additionally, this will only be reached when an analyzer has reported a diagnostic in the tree containing this syntaxRef, and hence we can expect the syntax nodes to have already been realized.